src/Interpreters/Context.cpp

```diff
@@ -556,7 +556,12 @@ struct ContextSharedPart : boost::noncopyable
             return;
 
         /// Need to flush the async insert queue before shutting down the database catalog
-        async_insert_queue.reset();
+        std::shared_ptr<AsynchronousInsertQueue> delete_async_insert_queue;
+        {
+            std::lock_guard lock(mutex);
+            delete_async_insert_queue = std::move(async_insert_queue);
+        }
+        delete_async_insert_queue.reset();
```

Comment:
Hm, and this is also wrong because other threads still can get a raw pointer from `getAsynchronousInsertQueue()` and continue using it after `reset()`